sklearn/base.py

```diff
@@ -556,6 +583,15 @@ def fit_transform(self, X, y=None, **fit_params):
             return self.fit(X, y, **fit_params).transform(X)
 
 
+class NonRectangularInputMixin:
+    """Mixin class for all estimators with non-rectangular input.
+
+    For now only vectorizers are relevant for this mixin.
+    """
+
+    n_features_in_ = None
```

Comment:
Since `n_features_in_` makes no sense for these, we decided to always set it to None.

Reply:
I still would rather not have a `*_` attribute for an unfitted object.

Reply:
Fair, I agree. I'll make it a property

ping @amueller who 	 might think otherwise

Reply:
"A foolish consistency is the hobgoblin of little minds"

What is the point of having estimators pass this? Adding code to make arbitrary checks pass seems like a bad idea.

Reply:
Should we just purely **not** have n_features_in_ for the vectorizers? (And remove the Mixin?)

@jnothman you suggested having it to None in the first place, WDYT?

Reply:
I'm okay having it removed, but then do we use a tag to indicate that input is not rectangular?

Reply:
> What is the point of having estimators pass this? Adding code to make arbitrary checks pass seems like a bad idea.

I thought we do care about API consistency a lot, and not just for the sake of passing the tests. Another way is to set it to None in fit, instead of as a class attribute. It'll still be one line of code.